pkg/services/ngalert/state/historian/annotation.go

```diff
@@ -30,45 +30,79 @@ func NewAnnotationHistorian(annotations annotations.Repository, dashboards dashb
+		logger.Debug("Alert state changed creating annotation", "newState", state.Formatted(), "oldState", state.PreviousFormatted())
 
-	labels = removePrivateLabels(labels)
-	annotationText := fmt.Sprintf("%s {%s} - %s", rule.Title, labels.String(), currentData.String())
+		labels := removePrivateLabels(state.Labels)
+		annotationText := fmt.Sprintf("%s {%s} - %s", rule.Title, labels.String(), state.Formatted())
+
+		item := annotations.Item{
+			AlertId:   rule.ID,
+			OrgId:     state.OrgID,
+			PrevState: state.PreviousFormatted(),
+			NewState:  state.Formatted(),
+			Text:      annotationText,
+			Epoch:     state.LastEvaluationTime.UnixNano() / int64(time.Millisecond),
+		}
 
-	item := &annotations.Item{
-		AlertId:   rule.ID,
-		OrgId:     rule.OrgID,
-		PrevState: previousData.String(),
-		NewState:  currentData.String(),
-		Text:      annotationText,
-		Epoch:     evaluatedAt.UnixNano() / int64(time.Millisecond),
+		items = append(items, item)
 	}
+	return items
+}
 
-	dashUid, ok := rule.Annotations[ngmodels.DashboardUIDAnnotation]
-	if ok {
-		panelUid := rule.Annotations[ngmodels.PanelIDAnnotation]
+// panelKey uniquely identifies a panel.
+type panelKey struct {
+	orgID   int64
+	dashUID string
+	panelID int64
+}
 
-		panelId, err := strconv.ParseInt(panelUid, 10, 64)
+// panelKey attempts to get the key of the panel attached to the given rule. Returns nil if the rule is not attached to a panel.
+func parsePanelKey(rule *ngmodels.AlertRule, logger log.Logger) *panelKey {
+	dashUID, ok := rule.Annotations[ngmodels.DashboardUIDAnnotation]
+	if ok {
+		panelAnno := rule.Annotations[ngmodels.PanelIDAnnotation]
+		panelID, err := strconv.ParseInt(panelAnno, 10, 64)
 		if err != nil {
-			logger.Error("Error parsing panelUID for alert annotation", "panelUID", panelUid, "error", err)
-			return
+			logger.Error("Error parsing panelUID for alert annotation", "actual", panelAnno, "error", err)
+			return nil
+		}
+		return &panelKey{
+			orgID:   rule.OrgID,
+			dashUID: dashUID,
+			panelID: panelID,
 		}
+	}
+	return nil
+}
 
-		dashID, err := h.dashboards.getID(ctx, rule.OrgID, dashUid)
+func (h *AnnotationStateHistorian) recordAnnotationsSync(ctx context.Context, panel *panelKey, annotations []annotations.Item, logger log.Logger) {
+	if panel != nil {
+		dashID, err := h.dashboards.getID(ctx, panel.orgID, panel.dashUID)
 		if err != nil {
-			logger.Error("Error getting dashboard for alert annotation", "dashboardUID", dashUid, "error", err)
+			logger.Error("Error getting dashboard for alert annotation", "dashboardUID", panel.dashUID, "error", err)
 			return
 		}
 
-		item.PanelId = panelId
-		item.DashboardId = dashID
+		for _, i := range annotations {
+			i.DashboardId = dashID
+			i.PanelId = panel.panelID
+		}
 	}
 
-	if err := h.annotations.Save(ctx, item); err != nil {
-		logger.Error("Error saving alert annotation", "error", err)
-		return
+	if err := h.annotations.SaveMany(ctx, annotations); err != nil {
```

Comment:
what do you think about adding a debug log after the save method? This method is async to the main state processing flow, and some indication might be helpful